feat(server): implement log rotation based on size and retention

Svecco opened the pull request:
Redirected from issue #46 , the progress:
- [x] implemented log rotation based on size and retention as the title;
- [x] implemented configurable attributes and imported breaking changes;
- [x] added units and integration test in logger.rs and integration mod;
- [x] added documentations and imported new dependencies, etc.

Plan:
*Around the next several weekends, before Mid February.*

- [x] ready for being merged;

hubcio commented:
I think changes in `core/message_bus` are not related to log retention, looks like merge problem.

spetz commented on core/server/src/log/logger.rs:
```diff
@@ -253,8 +255,48 @@ impl Logging {
-                tracing_appender::rolling::hourly(logs_path.clone(), IGGY_LOG_FILE_PREFIX);
+
+            if let Err(e) = std::fs::create_dir_all(&logs_path) {
+                tracing::warn!("Failed to create logs directory {:?}: {}", logs_path, e);
```
Let's import `warn` on top, next to `error, info`. Also, let's use string interpolation for all the messages, e.g. here and a few lines below.

spetz commented on core/server/src/log/logger.rs:
```diff
@@ -253,8 +255,48 @@ impl Logging {
+            }
+
+            // Check available disk space (at least 10MB)
+            let min_disk_space: u64 = 10 * 1024 * 1024; // 10MB
```
Let's put this value into const on top, something like `const MIN_DISK_SPACE_MB: u64 = 10`

spetz commented on core/server/src/log/logger.rs:
```diff
@@ -417,6 +457,160 @@ impl Logging {
+            let rotation_mutex_clone = Arc::clone(&rotation_mutex);
+            std::thread::spawn(move || {
+                loop {
+                    std::thread::sleep(Duration::from_secs(3600));
```
What was the reason behind 1h? IMHO this `duration` could be added to const on top.

spetz commented on core/server/src/log/logger.rs:
```diff
@@ -417,6 +457,160 @@ impl Logging {
+            }
+        }
+
+        if max_size_bytes > 0 {
```
To avoid the nested logic, this condition could be swapped with `if max_size_bytes == 0` + quick return from the function.

hubcio commented on core/server/src/log/logger.rs:
```diff
@@ -417,6 +457,160 @@ impl Logging {
+            let retention = config.retention.get_duration();
+            let rotation_mutex = Arc::new(Mutex::new(()));
+            let rotation_mutex_clone = Arc::clone(&rotation_mutex);
+            std::thread::spawn(move || {
```
this thread never terminates, please implement graceful shutdown

hubcio commented on core/server/src/log/logger.rs:
```diff
@@ -417,6 +457,160 @@ impl Logging {
+            let path = logs_path.to_path_buf();
+            let max_size = config.max_size.as_bytes_u64();
+            let retention = config.retention.get_duration();
+            let rotation_mutex = Arc::new(Mutex::new(()));
```
i don't understand what's the point of rotation mutex, only one thread is accessing `cleanup_log_files`

hubcio commented on core/server/src/configs/defaults.rs:
```diff
@@ -403,6 +403,8 @@ impl Default for LoggingConfig {
             file_enabled: SERVER_CONFIG.system.logging.file_enabled,
             max_size: SERVER_CONFIG.system.logging.max_size.parse().unwrap(),
+            max_total_size: SERVER_CONFIG.system.logging.max_total_size.parse().unwrap(),
+            rotation_check_interval: SERVER_CONFIG.system.logging.rotation_check_interval as u64,
```
please use IggyDuration, set default value to "1 h" and in `impl Validate for LoggingConfig` check if its higher than `1 s`

Svecco replied:
I have set the check interval to 1 hour and the upper limit of the rotation lifecycle to 7 days by default. Meanwhile, all default values have been streamlined to ensure they are determined by the integrated toml configuration.

Svecco replied:
> (see `validators.rs`)

I have added checks that cover certain special cases, such as an excessively high check frequency, among other scenarios.

hubcio commented on core/server/Cargo.toml:
```diff
@@ -68,10 +68,10 @@ error_set = { workspace = true }
 figlet-rs = { workspace = true }
 figment = { workspace = true }
 flume = { workspace = true }
+fs2 = "0.4"
```
please use full semver, we want somewhat reproducible builds

hubcio commented on core/server/src/log/logger.rs:
```diff
@@ -397,26 +444,199 @@ impl Logging {
+        max_files.clamp(1, 1000) as usize
+    }
+
+    fn _install_log_rotation_handler(&self, config: &LoggingConfig, logs_path: Option<&PathBuf>) {
```
remove prefix `_`

Svecco replied:
Alright, I will adhere to the relevant community guidelines.

hubcio commented on core/server/src/log/logger.rs:
```diff
@@ -397,26 +444,199 @@ impl Logging {
+                        break;
+                    }
+
+                    Self::cleanup_log_files(&path, retention, max_size);
```
you pass `max_size` here which is per-file size, but `cleanup_log_files` expects total limit

Svecco replied:
Fixed.

hubcio commented on core/server/src/log/logger.rs:
```diff
@@ -397,26 +444,199 @@ impl Logging {
+        }
+    }
+
+    fn read_log_files(
```
this function is somewhat incorrect - it looks for ANY file, perhaps you have to add something like `const IGGY_LOG_FILE_PREFIX` and filter the files based on that

Svecco replied:
Files currently being written to will no longer be accidentally deleted, achieved by marking the file list.

Svecco commented:
The log rotation test based on interval time and log volume is as console below:
<details> <summary>Console</summary>

```console
2026-01-01T09:03:36.553786Z  INFO shard-12 server::websocket::websocket_listener: WebSocket config: max_message_size: None, max_frame_size: None, accept_unmasked_frames: false
2026-01-01T09:03:37.523106Z DEBUG log-rotation server::log::logger: Starting log cleanup for directory: "local_data/logs", retention: 3600s, max_total_size: 4000000000 bytes, max_single_file_size: 512000000 bytes
2026-01-01T09:03:37.523237Z DEBUG log-rotation server::log::logger: Processed 4 log files from directory: "local_data/logs"
2026-01-01T09:03:37.523250Z DEBUG log-rotation server::log::logger: Mark old log file for remove: "local_data/logs/iggy-server.log.4"
2026-01-01T09:03:37.523260Z DEBUG log-rotation server::log::logger: Mark old log file for remove: "local_data/logs/iggy-server.log.1"
2026-01-01T09:03:37.523265Z DEBUG log-rotation server::log::logger: Mark old log file for remove: "local_data/logs/iggy-server.log.3"
2026-01-01T09:03:37.523270Z DEBUG log-rotation server::log::logger: Mark old log file for remove: "local_data/logs/iggy-server.log.2"
2026-01-01T09:03:37.530455Z DEBUG log-rotation server::log::logger: Removed log file: "local_data/logs/iggy-server.log.2"
2026-01-01T09:03:37.562069Z DEBUG log-rotation server::log::logger: Removed log file: "local_data/logs/iggy-server.log.3"
2026-01-01T09:03:37.569756Z DEBUG log-rotation server::log::logger: Removed log file: "local_data/logs/iggy-server.log.1"
2026-01-01T09:03:37.657651Z DEBUG log-rotation server::log::logger: Removed log file: "local_data/logs/iggy-server.log.4"
2026-01-01T09:03:37.657759Z  INFO log-rotation server::log::logger: Completed log cleanup for directory: "local_data/logs". Removed 4 files.
2026-01-01T09:03:38.657860Z DEBUG log-rotation server::log::logger: Starting log cleanup for directory: "local_data/logs", retention: 3600s, max_total_size: 4000000000 bytes, max_single_file_size: 512000000 bytes
2026-01-01T09:03:38.657958Z DEBUG log-rotation server::log::logger: Processed 0 log files from directory: "local_data/logs"
2026-01-01T09:03:38.657972Z  INFO log-rotation server::log::logger: Completed log cleanup for directory: "local_data/logs". Removed 0 files.
2026-01-01T09:03:39.658081Z DEBUG log-rotation server::log::logger: Starting log cleanup for directory: "local_data/logs", retention: 3600s, max_total_size: 4000000000 bytes, max_single_file_size: 512000000 bytes
2026-01-01T09:03:39.658171Z DEBUG log-rotation server::log::logger: Processed 0 log files from directory: "local_data/logs"
2026-01-01T09:03:39.658184Z  INFO log-rotation server::log::logger: Completed log cleanup for directory: "local_data/logs". Removed 0 files.
2026-01-01T09:03:40.658282Z DEBUG log-rotation server::log::logger: Starting log cleanup for directory: "local_data/logs", retention: 3600s, max_total_size: 4000000000 bytes, max_single_file_size: 512000000 bytes
2026-01-01T09:03:40.658376Z DEBUG log-rotation server::log::logger: Processed 0 log files from directory: "local_data/logs"
2026-01-01T09:03:40.658389Z  INFO log-rotation server::log::logger: Completed log cleanup for directory: "local_data/logs". Removed 0 files.
```

hubcio reviewed (commented):
I tested your code manually. In `server.toml` i changed settings to:
```toml
max_file_size = "1 MB"
max_total_size = "10 MB"
rotation_check_interval = "1 s"
```

then I rebulilt server and started it via:
`RUST_LOG=trace target/debug/iggy-server --with-default-root-credentials --fresh`

and in other shell i run benchmark to generate logs:
`target/debug/iggy-bench -T 1000MB pp -p 1 tcp`

result:
I can never see more than 2 files, it looks like retention is ignored and 10 MB of max total size is not reached.

can you write integration test that would test for this scenario? see how `TestServer` is used, in `integration` module (e.g. `core/integration/tests/server/specific.rs`)

Svecco commented:
Hi, I've completed some initial modifications and added integration tests.

I first manually observed the runtime behavior locally by running in the terminal: 
`RUST_LOG=trace cargo run --bin iggy-server -- --fresh --with-default-root-credentials` &&
`cargo run --bin iggy-bench -- -T 1000MB pp -p 1 quic`
I verified that the current log rotation behavior is working as expected by monitoring with 
`watch -n 0.1 "eza -la ~/Apache/iggy/local_data/logs"`

<details>
<summary>eza -la with 10:100</summary>

```Console
.rw-r--r--@ 2.4M svecco 11 Jan 13:10 iggy-server.log
.rw-r--r--@  10M svecco 11 Jan 13:04 iggy-server.log.1
.rw-r--r--@  10M svecco 11 Jan 13:04 iggy-server.log.2
.rw-r--r--@  10M svecco 11 Jan 13:04 iggy-server.log.3
.rw-r--r--@  10M svecco 11 Jan 13:04 iggy-server.log.4
.rw-r--r--@  10M svecco 11 Jan 13:04 iggy-server.log.5
.rw-r--r--@  10M svecco 11 Jan 13:04 iggy-server.log.6
.rw-r--r--@  10M svecco 11 Jan 13:03 iggy-server.log.7
.rw-r--r--@  10M svecco 11 Jan 13:03 iggy-server.log.8
.rw-r--r--@  10M svecco 11 Jan 13:03 iggy-server.log.9
```
</details>

Additionally, the result of the simple integration test `cargo test -p integration server::specific::log_rotation_should_launch -- --nocapture` execution seems fine, by generating sufficient logs by sending a large volume of messages to trigger the log rotation mechanism.

<details>
<summary>Console</summary>

```console
╭─ $ [svecco] ~/A/iggy git!(46-size-logs)
╰─ > cargo test -p integration server::specific::log_rotation_should_launch -- --nocapture
   Compiling libz-sys v1.1.23
   Compiling liblzma-sys v0.4.4
   Compiling hwlocality-sys v0.6.4
   Compiling libgit2-sys v0.18.3+1.9.2
   Compiling liblzma v0.4.5
   Compiling hwlocality v1.0.0-alpha.11
   Compiling compression-codecs v0.4.35
   Compiling async-compression v0.4.36
   Compiling async_zip v0.0.18
   Compiling git2 v0.20.3
   Compiling vergen-git2 v1.0.7
   Compiling server v0.6.1-edge.3 (/home/svecco/host/apache/iggy/core/server)
   Compiling integration v0.0.1 (/home/svecco/host/apache/iggy/core/integration)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 7.71s
     Running unittests src/lib.rs (target/debug/deps/integration-375a7f260e09ad62)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/mod.rs (target/debug/deps/mod-0f48c9a972f7dba1)

running 1 test
Log directory verified: local_data_136642763006894046097605999160554396417/logs
Succeeded Verified Log Rotation
test server::specific::log_rotation_should_launch ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 1190 filtered out; finished in 2.12s
```

Svecco commented:
(WIP) Here are new running status:

<details>
<summary>Cargo Test Console Results && 1MB:3MB Configured</summary>

```Console
╭─ $ [svecco] ~/A/iggy git!(46-size-logs)
╰─ > cargo test -p integration server::specific::log_rotation_should_launch -- --nocapture
    Finished `test` profile [optimized + debuginfo] target(s) in 0.16s
     Running unittests src/lib.rs (target/debug/deps/integration-da023a48286e4fe0)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/mod.rs (target/debug/deps/mod-6bd13926aa022770)

running 1 test

     Size <-> Path && server::specific::log_rotation_should_launch::nocapture_observer
  1000 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log.3
  1000 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log.2
  1000 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log.1
    39 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log


     Size <-> Path && server::specific::log_rotation_should_launch::nocapture_observer
  1000 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log.2
  1000 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log.1
    39 KB <-> local_data_318529726576567526029027823343213779713/logs/iggy-server.log

test server::specific::log_rotation_should_launch ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 1195 filtered out; finished in 2.33s
```
</details>

<details>
<summary>Server && Bench Side Console Commands</summary>

```Console
╭─ $ [svecco] ~/A/iggy git!(46-size-logs)
╰─ > RUST_LOG=trace cargo run --bin iggy-server -- --fresh --with-default-root-credentials

╭─ $ [svecco] ~/A/iggy git!(46-size-logs)
╰─ > cargo run --bin iggy-bench -- -T 50000MB pp -p 5 quic
```
</details>
<details>
<summary>watch -n 0.1 "eza -la ~/Apache/iggy/local_data/logs" && default configuration</summary>

```Console
	Every 0.1s: eza -la ~/Apache/iggy/local_data/logs
	orion: Sat Jan 17 22:31:29 2026

.rw-r--r--@ 365M svecco 17 Jan 22:31 iggy-server.log
.rw-r--r--@ 512M svecco 17 Jan 22:31 iggy-server.log.1
.rw-r--r--@ 512M svecco 17 Jan 22:31 iggy-server.log.2
.rw-r--r--@ 512M svecco 17 Jan 22:30 iggy-server.log.3
.rw-r--r--@ 512M svecco 17 Jan 22:30 iggy-server.log.4
.rw-r--r--@ 512M svecco 17 Jan 22:30 iggy-server.log.5
.rw-r--r--@ 512M svecco 17 Jan 22:30 iggy-server.log.6
.rw-r--r--@ 512M svecco 17 Jan 22:29 iggy-server.log.7

```
</details>
